python/src/spark_rapids_ml/core.py

```diff
@@ -660,6 +660,52 @@ def fitMultipleModels() -> List["_CumlModel"]:
         else:
             return super().fitMultiple(dataset, paramMaps)
 
+    def _try_stage_level_scheduling(self, rdd: RDD) -> RDD:
+        ss = _get_spark_session()
+        sc = ss.sparkContext
+
+        if ss.version < "3.4.0":
+            self.logger.warning(
+                "Stage level scheduling in spark-rapids-ml requires spark version 3.4.0+"
+            )
+            return rdd
+        elif _is_local(sc):
+            # Local mode doesn't support stage level scheduling
+            return rdd
+
+        executor_cores = sc.getConf().get("spark.executor.cores")
+        executor_gpu_amount = sc.getConf().get("spark.executor.resource.gpu.amount")
+
+        if executor_cores is None or executor_gpu_amount is None:
+            self.logger.warning(
+                "Stage level scheduling in spark-rapids-ml requires spark.executor.cores, "
+                "spark.executor.resource.gpu.amount to be set "
+            )
+            return rdd
+
+        task_gpu_amount = sc.getConf().get("spark.task.resource.gpu.amount")
+
+        if task_gpu_amount == executor_gpu_amount:
+            self.logger.warning(
+                f"The configuration of cores (exec = {executor_gpu_amount} task = {task_gpu_amount}, "
+                f"runnable tasks = 1) will result in wasted resources due to resource gpu limiting"
+                f"the number of runnable tasks per executor to: 1. Please adjust your configuration."
+            )
+            return rdd
+
+        from pyspark.resource.profile import ResourceProfileBuilder
+        from pyspark.resource.requests import TaskResourceRequests
+
+        # each training task requires cpu cores > total executor cores/2 which can
+        # ensure each training task be sent to different executor.
+        task_cores = (int(executor_cores) // 2) + 1
+        task_gpus = 0.1 if task_gpu_amount is None else float(task_gpu_amount)
```

Comment:
Would it suffice to set task_gpus = executor_gpu_amount?

Reply:
Hmm, `task_gpus = executor_gpu_amount` is going to limit only the training task running. If we set it to a fraction, we can still allow other ETL tasks to run alongside training tasks.

Reply:
You mean from other Spark jobs running in parallel?   I suppose it could also allow other tasks on the GPU as well.   Do we want to allow sharing of GPU with another Spark job?   Seems a recipe for GPU OOMs.

Reply:
Really good point. so do you suggest setting task_gpus = executor_gpu_amount? @eordentlich

Reply:
Would task_gpus=1  actually be the ideal choice for now?  That would address @lijinf2 's concern as well.   
But if improved performance can be demonstrated for some use case involving parallel cpu jobs with gpu jobs, maybe we should allow it (and just resolve OOMs via resource configs).
Also, the fit stage scheduling probably should not lower task_gpus below the original setting so that it can be pinned to 1 if needed.

Reply:
Good. new commit has changed it.